src/Owin.Security.Providers.CanvasLMS/CanvasAuthenticationHandler.cs

```diff
@@ -228,5 +231,83 @@ protected override Task ApplyResponseChallengeAsync()
 
             return context.IsRequestCompleted;
         }
+
+        private async Task<object> RequestToken(string grantType, string grantToken, string grantTokenParameterName = null)
+        {
+            var requestPrefix = Request.Scheme + "://" + Request.Host;
+            var redirectUri = requestPrefix + Request.PathBase + Options.CallbackPath;
+
+            // Build up the body for the token request
+            var body = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>(grantTokenParameterName ?? grantType, grantToken),
+                new KeyValuePair<string, string>("grant_type", grantType),
+                new KeyValuePair<string, string>("client_id", Options.ClientId),
+                new KeyValuePair<string, string>("redirect_uri", redirectUri)
+            };
+
+            // Request the token
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.EndpointBase + Options.Endpoints.TokenPath);
+            requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Basic",
+                new Base64TextEncoder().Encode(Encoding.ASCII.GetBytes(Options.ClientId + ":" + Options.ClientSecret)));
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            requestMessage.Content = new FormUrlEncodedContent(body);
+
+            var tokenResponse = await _httpClient.SendAsync(requestMessage);
+            tokenResponse.EnsureSuccessStatusCode();
+
+            var tokenResponseContent = await tokenResponse.Content.ReadAsStringAsync();
+            return JObject.Parse(tokenResponseContent);
+        }
+
+        private async Task<CanvasAuthenticatedContext> Authenticate(dynamic response, AuthenticationProperties properties = null, string refreshToken = null)
```

Comment:
Again, copy/pasta. If I have to touch this again, I'll probably layer in some types.